ipfs-8-to-9/migration/migration.go

```diff
@@ -0,0 +1,242 @@
+
+const backupFile = "8-to-9-cids.txt"
+
+// Migration implements the migration described above.
+type Migration struct{}
+
+// Versions returns the current version string for this migration.
+func (m Migration) Versions() string {
+	return "8-to-9"
+}
+
+// Reversible returns false. This migration cannot be reverted, as we do not
+// know which raw hashes were actually CIDv1s. However, things should work all
+// the same as they will be treated as CIDv0s in old versions anyways.
+func (m Migration) Reversible() bool {
+	return false
+}
+
+// lock the repo
+func (m Migration) lock(opts migrate.Options) (io.Closer, error) {
+	log.VLog("locking repo at %q", opts.Path)
+	return lock.Lock2(opts.Path)
+}
+
+// open the repo
+func (m Migration) open(opts migrate.Options) (ds.Batching, error) {
+	log.VLog("  - loading repo configurations")
+	plugins, err := loader.NewPluginLoader(opts.Path)
+	if err != nil {
+		return nil, fmt.Errorf("error loading plugins: %s", err)
+	}
+
+	if err := plugins.Initialize(); err != nil {
+		return nil, fmt.Errorf("error initializing plugins: %s", err)
+	}
+
+	if err := plugins.Inject(); err != nil {
+		return nil, fmt.Errorf("error injecting plugins: %s", err)
+	}
+
+	cfg, err := fsrepo.ConfigAt(opts.Path)
+	if err != nil {
+		return nil, err
+	}
+
+	dsc, err := fsrepo.AnyDatastoreConfig(cfg.Datastore.Spec)
+	if err != nil {
+		return nil, err
+	}
+
+	return dsc.Create(opts.Path)
+}
+
+// Apply runs the migration and writes a log file that can be used by Revert.
+func (m Migration) Apply(opts migrate.Options) error {
+	log.Verbose = opts.Verbose
+	log.Log("applying %s repo migration", m.Versions())
+
+	lk, err := m.lock(opts)
+	if err != nil {
+		return err
+	}
+	defer lk.Close()
+
+	repo := mfsr.RepoPath(opts.Path)
+
+	log.VLog("  - verifying version is '8'")
+	if err := repo.CheckVersion("8"); err != nil {
+		return err
+	}
+
+	dstore, err := m.open(opts)
+	if err != nil {
+		return err
+	}
+	defer dstore.Close()
+
+	// Assuming the user has not modified the blocks namespace
+	blocks := namespace.Wrap(dstore, ds.NewKey("/blocks"))
+
+	log.VLog("  - starting CIDv1 to raw multihash block migration")
+
+	// Prepare backing up of CIDs
+	backupPath := filepath.Join(opts.Path, backupFile)
+	log.VLog("  - backup file will be written to %s", backupPath)
+	_, err = os.Stat(backupPath)
+	if err != nil {
+		if !os.IsNotExist(err) {
+			log.Error(err)
+			return err
+		}
+	} else { // backup file exists
+		log.Log("WARN: backup file %s already exists. CIDs-Multihash pairs will be appended", backupPath)
+	}
+
+	// If it exists, append to it.
+	f, err := os.OpenFile(backupPath, os.O_WRONLY|os.O_APPEND|os.O_CREATE, 0644)
+	if err != nil {
+		log.Error(err)
+		return err
+	}
+	defer f.Close()
+	buf := bufio.NewWriter(f)
+	defer buf.Flush()
+
+	swapCh := make(chan Swap, 1000)
+
+	var wg sync.WaitGroup
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for sw := range swapCh {
+			fmt.Fprint(buf, sw.Old.String()+","+sw.New.String()+"\n")
```

Comment:
To be safe, I think we need to write this _first_. We might want to:

1. Walk the datastore, writing down every CIDv1 we come across.
2. Do it again, doing the swaps.

Reply:
I.e., write ahead logging: write what we're going to do (what we're going to migrate), do it (swap), then say we did it (update the repo version).

Reply:
Note: we also don't need to record old/new. We can derive the multihash from the CID (and cut the filesize in half).

Reply:
> To be safe, I think we need to write this _first_. We might want to:
> 
>     1. Walk the datastore, writing down every CIDv1 we come across.
> 
>     2. Do it again, doing the swaps.

Ok, done.

> Note: we also don't need to record old/new. We can derive the multihash from the CID (and cut the filesize in half).

Ok, done this too.